### DISCUSSION TITLE
Difference between `utils/readline/` and https://github.com/lmorg/readline

### DISCUSSION THREAD

**Discussion by itaranto** [lmorg/murex#891] [Q&A] [answered]
Hello,

This is not a `murex` question specifically but I got interested in the `readline` library that apparently got extracted from this project.

I have a couple of questions:
- How different are the `readline` implementations from `utils/readline/` and https://github.com/lmorg/readline?
- Is https://github.com/lmorg/readline still maintained?
- Why https://github.com/lmorg/readline isn't a dependency of https://github.com/lmorg/murex?

**Comment by lmorg** [comment, marked as answer]
Good questions.

> How different are the readline implementations from utils/readline/ and https://github.com/lmorg/readline?

`utils/readline` has quite a number of new features and bug fixes:

- Windows is now fully supported in `utils/readline` whereas there are edge case bugs with syscalls and ANSI escape sequences with with `lmorg/readline` when in Windows
- new features like event handling and preview panes
- improved performance
- a few bug fixes so edge cases shouldn't result in a panic

However the API is compatible so you "should" be able to use `utils/readline` as a drop in replacement for `lmorg/readline`

> Is https://github.com/lmorg/readline still maintained?

Not really. Or at least not as an independent repository. There were a few other readline libraries that have emerged since so I got lazy backporting updates from `utils/readline` to `lmorg/readline`.

> Why https://github.com/lmorg/readline isn't a dependency of https://github.com/lmorg/murex?

Originally it was. But this was back when `$GOPATH` was required. `go mod` made it harder to develop readline in parallel with Murex so I got lazy and just copied the readline project into the Murex directory.

The intention was always to split the projects again, and I've purposely kept `utils/readline` code non-reliant on any on anything Murex so I can do that at a later date. I've just been exceptionally lazy since there's always a backlog of more impactful tasks to work on Murex.

I'm always open to suggestions and feedback though.

**Comment by lmorg** [comment]
@itaranto `utils/readline` has now been merged back into `github.com/lmorg/readline` and going forward, that will be the home for `readline` and thus Murex will import it from there rather than an internal `utils` path.
